### PR TITLE
IODEMO: rework buffers management

### PR DESCRIPTION
**Pull request by evgeny-leksikov**

### CODE REVIEW

**Comment by evgeny-leksikov** [comment]
@yosefe pls take a look

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
         XFER_TYPE_RECV
     } xfer_type_t;
 
+    class BufferChunk {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: maybe just "class Buffer"

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
...
+    public:
+        BufferIov(size_t size, MemoryPool<BufferIov> *pool) : _pool(*pool) {
+            _iov.reserve(size);
+        }
+
+        size_t size() const {
+            return _iov.size();
+        }
+
+        void init(size_t data_size, MemoryPool<BufferChunk> &chunk_pool,
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: why here by & and line 210 by *

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
...
+                  uint32_t sn, bool fill) {
+            assert(_iov.empty());
+
+            BufferChunk *chunk = chunk_pool.get();
+            _iov.resize(get_chunk_cnt(data_size, chunk->_capacity));
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:223]: could be malloc on fast path?

File: test/apps/iodemo/io_demo.cc [hunk 4]
```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
...
+            _size = size;
+        }
+
+        inline size_t size() const {
+            return _size;
+        }
+
+        ~BufferChunk() {
+            free(_buffer);
+        }
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: move this right after ctor?

File: test/apps/iodemo/io_demo.cc [hunk 5]
```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
...
+            assert(_iov[0]->size() >= (sizeof(chk_hdr_t) + sizeof(tr_hdr_t)));
+            return reinterpret_cast<chk_hdr_t*>(_iov[0]->buffer());
+        }
+
+        tr_hdr_t* tr_hdr() const {
+            return reinterpret_cast<tr_hdr_t*>(chk_hdr() + 1);
+        }
+
+    private:
+        std::vector<BufferChunk *> _iov;
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: can remove space before *

File: test/apps/iodemo/io_demo.cc [hunk 6]
```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
...
+
+    private:
+        void *                    _buffer;
+        size_t                    _size;
+        MemoryPool<BufferChunk> & _pool;
+    };
+
+    class BufferIov {
+    public:
+        BufferIov(size_t size, MemoryPool<BufferIov> *pool) : _pool(*pool) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: imo need to always pass pool by & in all places for consistency

File: test/apps/iodemo/io_demo.cc [hunk 7]
```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
...
+
+        uint16_t chk_sum() const {
+            return chk_hdr()->chk_sum;
+        };
+
+        uint32_t sn() const {
+            return tr_hdr()->sn;
+        }
+
+        inline uint16_t calc_chk_sum() const {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: calc_checksum

File: test/apps/iodemo/io_demo.cc [hunk 8]
```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
...
+            uint16_t chk_sum = ucs_crc16(tr_hdr(), (uintptr_t)tr_hdr() -
+                                                   (uintptr_t)_iov[0]->buffer());
+
+            for (size_t i = 1; i < _iov.size(); ++i) {
+                chk_sum ^= ucs_crc16(_iov[i]->buffer(), _iov[i]->size());
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:263]: maybe add crc16 function to Buffer class

**Comment by yosefe** [comment]
pls asl ocheck the PR does not reduce performance with "-d 65536 -w 32" for both read and write

File: test/apps/iodemo/io_demo.cc (hunk 3 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:223]: nope, ctor does `reserve()`

File: test/apps/iodemo/io_demo.cc (hunk 8 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:263]: Buffer does not know about possible header in first chunk, this is why it's handled in Iov

Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:263]: i mean, each buffer will compute its check sum independently and then xor all in this func
BTW, why need special case for buffer 0, and not just init chk_sum to 0?

File: test/apps/iodemo/io_demo.cc (hunk 3 above)
Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:223]: and vector will not try to realloc or release mem after that?

Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:223]: yes, but only if it's reserved correctly and capacity is not increased by resize()
(need to fix ctor param on initialization)

File: test/apps/iodemo/io_demo.cc (hunk 2 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc]: elem ctor is part of MemoryPool API, will fix it

File: test/apps/iodemo/io_demo.cc (hunk 8 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:263]: crc will be stored in header of first buffer, buffers is not aware about this and result crc will be wrong
> BTW, why need special case for buffer 0, and not just init chk_sum to 0?

exactly by this reason, crc of first buffer must be calculated with header offset

Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:263]: does it include sn?

Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:263]: yes, tr_hdr() contains this, initialized by `fill_data` on sender side or received with data on receiver side

**Comment by evgeny-leksikov** [comment]
> pls asl ocheck the PR does not reduce performance with "-d 65536 -w 32" for both read and write

perf is OK

File: test/apps/iodemo/io_demo.cc [hunk 9]
```diff
@@ -88,6 +88,10 @@ class MemoryPool {
         if (_freelist.empty()) {
             item = new T(_buffer_size, *this);
             _num_allocated++;
+            if (_verbose && ((_num_allocated % 100) == 0)) {
+                UcxLog("MEM_POOL", _verbose) << "pool " << this
+                                             << ": allocated " << _num_allocated;
+            }
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: print this along with printing bandwidth

File: test/apps/iodemo/io_demo.cc [hunk 10]
```diff
@@ -262,94 +422,56 @@ class P2pDemoCommon : public UcxContext {
...
-        if (sn != peer_sn) {
+        if (sn != iov.sn()) {
             pass = false;
-            LOG << "detected transaction mismatch " << sn << " != "
-                << peer_sn;
+            LOG << "detected transaction mismatch " << sn << " != " << iov.sn();
         }
 
-        if (chk_sum != peer_chk_sum) {
+        uint16_t recalc_chksum = iov.calc_chksum();
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:453]: pls add comment to explain

File: test/apps/iodemo/io_demo.cc (hunk 9 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc]: but then we can control only data buffers pool, this print allows to catch any pool leak

Comment (yosefe) [reply test/apps/iodemo/io_demo.cc]: print all pools status in same line where bw is printed:
"data-pool: %u/%u cb-pool: %u/%u", inuse/total-allocated, inuse/total-allocated

Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc]: ok, but only client prints bandwidth

Comment (yosefe) [reply test/apps/iodemo/io_demo.cc]: good point
let's move the print to next PR to polish it later

Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc]: in addition, we report perf when window is done, so inuse == 0

File: test/apps/iodemo/io_demo.cc (hunk 10 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:453]: done

File: test/apps/iodemo/io_demo.cc (hunk 9 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc]: reverted

File: test/apps/iodemo/io_demo.cc [hunk 11]
```diff
@@ -449,6 +449,7 @@ class P2pDemoCommon : public UcxContext {
             LOG << "detected transaction mismatch " << sn << " != " << iov.sn();
         }
 
+        /* recalc check sum of all frigments and compare to stored value */
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: typo: fragments

**Review by yosefe — COMMENTED** [review]
pls squash after typo fix

**Review by yosefe — APPROVED** [review]

**Pull request merged**